torch/profiler/profiler.py

```diff
@@ -0,0 +1,191 @@
+import torch.autograd.profiler as prof
+from torch.autograd import ProfilerActivity
+
+from typing import Callable, Iterable, Optional
+
+class EnablePred(object):
+    """
+    EnablePred describes on which steps profiler is active:
+    - profiler starts in inactive state and stays in inactive state for the first 'wait' steps
+    - profiler then enters a warmup state and stays in this state for the next 'warmup' steps
+    - profiler then starts actively tracing/collecting stats for the next 'active' steps
+    - after this, profiler returns to the inactive state and cycle repeats
+
+    In case output_fn is specified, it is called every time the trace is ready
+    """
+    class Action(object):
+        START_WARMUP = 0
+        START_TRACE = 1
+        STOP_TRACE = 2
+
+    class State(object):
+        INACTIVE = 0
+        WARMUP = 1
+        ACTIVE = 2
+
+    def __init__(self, wait: int, warmup: int, active: int, output_fn: Optional[Callable[[prof.profile], None]]):
+        assert wait >= 0 and warmup >= 0 and active > 0
+        if warmup == 0:
+            print("Warning: profiler won't be using a warmup, which can skew profiler results")
+        self.wait = wait
+        self.warmup = warmup
+        self.active = active
+        self.output_fn = output_fn
+
+        def active_active_fn(step):
```

Comment:
The code here is very challenging to read and looks almost like a state machine. I guess I was expecting something simpler (like taking the modulus of the current iteration or epoch). Is this additional complexity really necessary?

Reply:
it is indeed a state machine, yes

Reply:
Can you elaborate on why that's necessary (vs taking a modulus)?

Reply:
I can e.g. give an example showing that it's not so simple.
So you take the modulus and you know your current state, is it enough?
You still need to know what was the previous state and what where in the sequence you are.
If the current state is active, and the previous is warmup - action is start_trace
If the current state is active, and the previous is inactive - action is start_warmup, followed by start_trace
If the current state is active, and the previous state is active, does it mean we don't need to do anything?
It depends, consider (wait=0, warmup=0, active=2) - sometimes we don't do anything, sometimes we do stop, save, warmup, start.

So you still have to consider a lot of possibilities and edge cases, then in that case, why not just encode states and actions explicitly in the map? The original version of the code that didn't use it was way less readable imo.

Reply:
What actually happens when wait=0, warmup=0 and active=2? I would expect the following pattern:

ACTIVE ACTIVE ACTIVE ACTIVE ACTIVE...

Is this not correct?

Reply:
"wait=0, warmup=0 and active=2" means don't wait, don't do warm up, trace for two iterations

Reply:
then the sequence repeats, and we provide access to the traces

Reply:
Oh interesting. So it's more correctly:

ACTIVE ACTIVE REPORT ACTIVE ACTIVE REPORT

?

Reply:
"REPORT" is implied at the every end of enable_pred cycle, will make sure to make it clear from the docs